### ISSUE TITLE
Is a standard needed?

### ISSUE THREAD

**Issue by DaveLiddament** [DaveLiddament/php-generics-standard#2]
**Reasons for formalising a standard**

1. It *might* increase uptake. The lack of a standard is the reason cited in PHPStorm's 2020.3 [release announcement](https://blog.jetbrains.com/phpstorm/2020/12/phpstorm-2020-3-release/#psalm_and_phpstan_support) for not fully supporting `@template`:
> We believe that support for generics is an advanced feature that lacks a proper specification and has many edge cases. Yet, we have decided to implement basic support for the @template construct based on the Psalm syntax, to see how it goes.
2. Library and framework maintainers might be happier to have the generics annotations in their code if there was a standard.

**Reasons for aborting this project**
1. Some libraries already support generics, so clearly a formal standard is not needed. 
2. Those who want the power of generics have it already. If they make use of a library that doesn't have the generics annotations then stubs and plugins exist, or can be written, to provide support with static analysers
3. A lot of work!

**Comment by DaveLiddament** [comment]
Another reason against is that PhpStorm does support [use of PHPStan](https://www.jetbrains.com/help/phpstorm/using-phpstan.html) and [Psalm](https://www.jetbrains.com/help/phpstorm/using-psalm.html).

This might be good enough for the majority of developers.

**Comment by PatchRanger** [comment]
# Preamble
In the context of a recent discussion about generics and how to achieve them: https://thephp.foundation/blog/2025/08/05/compile-generics/ .
In my opinion, the attribute approach to the generics is the simplest, most transparent and most declarative.
I will try to explain below why a single standard is needed, why attributes are better than PhpDoc-comments and how to get them without the need for enormous effort on anyone's part.

# Minimum option
Create an RFC with the addition of the #PhpDoc-attribute, which does the only thing: adds **literally the same** (this is important) corresponding PhpDoc-comment.

## Why, what's the point if there are already PhpDoc in the form of comments?
We have already seen a similar discussion in the past: why do we need attributes at all if there are comments?
And the answer is known: comments are not code, but attributes are code.
Here are the reasons:
1) Validation at the time of writing. It prevents creation of misconfigured/misspelled PhpDocs.
2) The possibility to add additional logic of processing these attributes.
For example, the Deprecated attribute triggers warning precisely because it is part of the code markup, and not a comment on the code.
3) Working with them is faster than with PhpDoc-comments: attributes are available through reflection, there is no need to parse comments and process typos and errors.

## So, why those attributes are needed to be in the core, why "as is" (just PhpDoc-comments) is not enough?
Standard is like a self-fulfilling prophecy: the more widely these attributes are used, the more widely they are used.
For some, it's just enough that the attributes look prettier than the PhpDoc-comments.
Someone will be tempted by the opportunity to ground their own logic on that shiny new attributes.
Static analyzers can be bribed by the fact that attributes are faster to work with than comments (as they require parsing which has to handle possible typos and therefore could be not so easy and fast), which means it is advantageous to add support for these attributes: it is relatively easy, because they are literally like PhpDocs that are already supported.
And as a result, it may turn out that such a simple combination of introductory elements is enough to ensure the increasing spread and adoption of new attributes:
- PHPDoc-attribute in the PHP core (which simply adds the corresponding PHPDoc-comment, which means it "just works" out of the box);
- the Rector rule, which turns the PHPDoc-comment into the PHPDoc-attribute (because we can and it's beautiful).

That is, I suggest not trying to invent some kind of innovation and bring it down from above, to plant it with fire and sword, but to establish a process of generalizing the practice of using PhpDoc-comments and "canonizing" well-proven and well-known PhpDocs by such a tricky PHPDoc-attribute.

## Why is so important to make this new attribute to work and look literally the same as PhpDoc-comment it refers to?
1) It will be relatively easy to automatically convert (for example, through the Rector) a huge bunch of existing code from the PhpDoc-comments into these PhpDoc-attributes (thereby ensuring their validity). Looking ahead: it's also true about 4 attributes of Dave (see the extended version below).

2) This will allow us to implement support for convenient attributes in the form of an "alias" to the corresponding PHPDoc attribute.

```
// Below is just a pseudocode to illustrate the point that it would be rather simple to add new PhpDoc-attributes "that just work" at such basis.
#Template = #PhpDoc('template')
#Type = #PhpDoc('var') | #PhpDoc('param') | #PhpDoc('return')
```
And this opens the way for an Extended version.

# Extended version
Minimal version + Supplement the RFC with a proposal to "canonize" in the form of attributes those 4 attributes that Dave proposed 4 years ago.

Thereby, I suggest combining the advantages of both worlds:
- The flexibility and simplicity of erasable generics in the form of community-provided PhpDocs.
- Standardization and strictness (including "compile-time" validation) of attributes from the PHP core team.

And that's how:
1) Let thousands of flowers bloom in the form of various PhpDoc-comments, each library has its own, it's wonderful, it's search and creativity.
This is an unobtrusive way to extend the functionality of the language.

2) On the other hand, the PHP community always has the opportunity at any time to raise the issue of "canonizing" of any of the successful PhpDocs in any convenient/pleasant/pragmatic way by adding the corresponding attribute to the language.
Exactly as it is happening right now thanks to Dave's efforts: 4 years ago he proposed including 4 attributes in the language, and right now I'm offering a simple and pragmatic way how exactly to do it.
This is a way to consolidate successful and widespread findings, make them part of the general syntax of the language, formalize and subordinate them to rules, and calm the discord.

# TL;DR
In total, I propose to create an RFC that adds support for generics to the core of the PHP language, with 5 votings about adding 5 attributes.:
1) #[PhpDoc]: fire-starter and game-changer, without it my proposal is indistinguishable from the original one, so a vote against it means votes against everyone else.
Required to be implenebted in the form of adding a comment with the corresponding PHPDoc-comment.
2) #[Type]: with an implementation in the form of #[PHPDoc('var')], #[PHPDoc('param')] and #[PHPDoc('return')] respectively.
The essence and the meaning of this attribute is simply to add PhpDoc-comment that is already working out-of-the-box.
3) #[Template]: with an implementation in the form of #[PHPDoc('template')].
The essence and the meaning of this attribute is simply to add PhpDoc-comment that is already working out-of-the-box.
4) #[Extend]: with an implementation in the form of #[PHPDoc('template-extends')].
The essence and the meaning of this attribute is simply to add PhpDoc-comment that is already working out-of-the-box.
In my opinion, it leads to false thoughts about class inheritance, rather than template, I would suggest #[TemplateExtends]: unambiguous and literally reproduces the well-known code.
4) #[Implement]: with an implementation in the form of #[PHPDoc('template-implements')].
The essence and the meaning of this attribute is simply to add PhpDoc-comment that is already working out-of-the-box.
In my opinion, it leads to false thoughts about the implementation of the interface by a class, and not by a template, I would suggest #[TemplateImplements]: unambiguous and literally reproduces the well-known code.

This is a way not to destroy the good that has already matured through the efforts of the community. And at the same time, it is a foundation for a bright future, in which it is clear how to move: to build logic on top of this attribute-based markup and new attributes based on the best experience of PhpDocs.
Thank you for your attention.
I will be glad for any responses.

I hope it won't be too immodest of me to invite into the discussion those who might be interested in it:
- @DaveLiddament (as author of original effort to attribute-based generics)
- @Crell and @Girgias (as authors of that exciting post about generics)
- @danog (as maintainer of Psalm)
- @ondrejmirtes (as maintainer of PHPStan)

**Comment by Girgias** [comment]
If you want to include relevant people of the ecosystem, you should also mention @azjezz as maintainer of Mago, @SerheyDolgushev as one of the active Phan maintainer, @dseguy / @exakat as maintainer of Exakat, and probably look what other static analysis tools there are.

As my own opinion, while standardizing an attribute approach can definitely make sense, I don't agree that this should be in Core, as there is no benefit for core to do this approach, the main reason I started working on what was initially Associated Types (which then became abstract generics) was so that I could *use* them with the new Container interfaces that I proposed previously in https://wiki.php.net/rfc/container-offset-behaviour, attributes do not allow me this.
Moreover, I have a working implementation for *just* interfaces.

I haven't caught up on all the discussion on internals about that post because it dropped days before feature freeze landed, and I am currently on holiday. But from what people have indicated to me, the main complaint seems to be about "abstract classes" being supported with no syntax for `new ClassWithGenerics<T, U>` which has good reasons for not existing at the time being. As such I might just drop the whole abstract class thing anyway until people can agree about what the hell should be done and only propose generics in interfaces, as this would already net improvements for core interfaces, and even in my own PHP code that I write, have *just* generics in interfaces solves some annoying problems already *for me*.

This sort of standardization is perfect for the realm of FIG and a PSR rather than core.

**Comment by ondrejmirtes** [comment]
I'm absolutely against this proposal. There's nothing to gain from going from `/**` to `#[`. PHPDoc-PHPDocs are already validated and used by static analysis tools, they can't be considered "just a comment" as opposed to "proper code".

Additionally, even with your proposal of Attribute-PHPDocs, you'd still have types as plain strings, still needing library like phpstan/phpdoc-parser.

PHPDoc-PHPDocs are already well supported across the ecosystem. The move from PHPDoc-PHPDocs to Attribute-PHPDocs would bring a lot of work for negligible gains.

For as long as I remember, the trend in PHP language development was that features/ideas first introduced by PHPDocs eventually made it to native syntax. Things like return typehints, scalar typehints, `void`, `never`, nullables etc. I expect this trend to continue.

Attribute-PHPDocs would be just a stopgap that we as the community can skip.

There's a hacked-together prototype of "types in attributes" that you can look at (https://github.com/php-static-analysis/attributes) and maybe understand why I'm not a fan of that. Instead of very clean PHPDocs everyone is used to, you have types in attributes and every attribute ends with a bunch of inelegant characters: `')]`

**Comment by azjezz** [comment]
> In my opinion, the attribute approach to the generics is the simplest, most transparent and most declarative.

This is not a good idea. If attribute generics are just for static analysis tools, then there's no functional difference between that and PHPDoc.

If this is for the PHP engine itself, it's also not a good idea. There aren't syntax problems that require us to rely on attributes. `<T1, T2, ..Tn>` syntax is possible for definitions and for `implements`/`extends`/`use`. For expressions, turbofish syntax (e.g., `parse::<int>($input)`) can resolve the syntax ambiguity.

> Validation at the time of writing. It prevents creation of misconfigured/misspelled PhpDocs.

Linters can already catch this. Tools like PHPStan, Psalm, and Mago all report invalid PHPDoc annotations.

> The possibility to add additional logic of processing these attributes. For example, the Deprecated attribute triggers warning precisely because it is part of the code markup, and not a comment on the code.

The `#[Deprecated]` attribute is a directive for the PHP engine that doesn't affect the runtime behavior of the class. This is different from generics, which fundamentally alter type interactions.

> Working with them is faster than with PhpDoc-comments: attributes are available through reflection, there is no need to parse comments and process typos and errors.

Not necessarily. Most static analysis tools have to parse the file anyway. Parsing PHPDoc doesn't significantly slow things down. Some static analysis tools use libraries like `better-reflection`, which re-parses the file, while others like Mago parse everything themselves. Having to parse a PHPDoc comment on top of the file isn't a real performance bottleneck for us.

> Standard is like a self-fulfilling prophecy: the more widely these attributes are used, the more widely they are used.

I disagree. If standardization for static analysis is the goal, then the maintainers of the static analysis tools should agree on the specifications. This shouldn't be decided by PHP-src maintainers who won't be implementing or maintaining the static analysis-side logic. Unless the plan is to add full, native generics support in PHP-src itself, there is no reason for php-src to dictate how Psalm, PHPStan, or Mago should handle generics.

**Comment by Crell** [comment]
There would be advantages to a standard "Extended type" system in attributes rather than docblocks.  In particular, it makes it easier for everyone else to parse them, without relying on PHPStan/Psalm.  For example, in Crell/Serde, I need to know the key and value types of arrays.  That's basically generic information.  To avoid having to pull in a docblock parser, I added my own attribute to contain that information, which gets parsed "for me" by Reflection.  But it does mean the information must be double-entered, once in my attribute and once in a normal docblock for SA tools.  It would be helpful if there were a standard attribute that we could both use.

However!  That attribute *would not be useful if it used the same type signature as current docblocks*, because that would still involve string parsing on my part.  That's the thing to avoid.  It would provide me no value to have an attribute `#[Type('string[]')]`, because then I'd still need to parse the type string to figure out that it's an array of strings, which could also be spelled `#[Type('array<string>')]`, so I have to be able to parse both.  I'm just not going to bother.

What would be beneficial instead is `#[Sequence(Point::class)]`, or `#[Dictionary('string', Point::class)]`, which gives me the type information pre-separated, and distinguishes sequences from dictionaries (something any solution MUST do).  Which is exactly what Serde does now.  (Actually it uses an enum for the key of a dict, since there's only 2 legal possibilities.)

So a direct port of the current syntax, which is in many cases quite cumbersome and annoying, would not only not provide much value, I'd argue it would be a net negative.

As far as where such should live:

* Attributes should only be in core if core is going to use them.  If they're just a convention, that's not php-src's job.
* Standardizing conventions is PHP-FIG's job.  FIG actually just chartered a new Working Group for a PER to provide some standardized attributes.  So that would be where any such discussion should take place.  (Disclosure: I am a member of the working group, though Andreas is the Editor.)

However!  One of the drums I regularly beat in FIG is that standards that are unused are not helpful, so whether FIG is standardizing existing conventions or developing some new agreement, it needs buy-in.  It sounds like the most important players in the space are vehemently against using attributes (rightly or wrongly), so I'd be skeptical of the FIG WG including generics attributes if the major SA tools have explicitly said they will ignore them.  (Unless the SA tools would be willing to support attributes that have FIG backing?)

I have long argued that erased generics are of minimal value to PHP, and to the extent they are useful, they are already adequately provided by SA tools.  Anything in-core should have impact in-core.  Be that full generics, limited partial generics, associated types, or whatever else: Core itself should not have do-nothing features.

**Comment by azjezz** [comment]
> `#[Type('string[]')]`, because then I'd still need to parse the type string to figure out that it's an array of strings, which could also be spelled `#[Type('array<string>')]`
> `#[Sequence(Point::class)]`, or `#[Dictionary('string', Point::class)]`

This is not useful for generics. For example:

```php
<?php

/**
 * @template K of array-key
 * @template V
 */
class Collection {
    /**
     * @param K $k
     * @return V
     */
    public function get(string|int $k): mixed { /* ... */ }
}
```

How would this be implemented using the proposed attributes? What are `Sequence` and `Dictionary`? Are they aliases for `list<V>` and `array<K, V>`? If not, and they are collection-like types, why are they accepting class-strings in their constructors and why are they themselves attributes?

This is generally not possible to implement using attributes in this manner.

Templates are not just type strings to be parsed; you also need to carry a type context around to resolve what names refer to. For instance, `K` in the `get` method's docblock could be a class named `K`, a type alias imported from somewhere else, a template parameter from the current class, a template parameter from the current method, or it could be an invalid reference that doesn't exist.

For serialization/deserialization, if you wish to support generics, you must use a tool like PHPStan that can perform this kind of complex type resolution for you.

---

I'm sorry if this comes across as harsh, but this whole attribute proposal makes zero sense to me. The path forward should be one of the following:

1. If php-src wishes to support generics, it should add real, native syntax.
2. If php-src wants to help static analysis tools, the answer is still to add real syntax. It could be parsed and made available via reflection without initial runtime checks. The runtime component could be added later.
3. If the goal is simply a standard for SA tools, we already have a de facto one: PHPDoc. The minor differences between tools should be decided upon by their respective maintainers.

**Comment by Crell** [comment]
@azjezz To be clear, a single "Collection" class that doesn't differentiate whether it is a sequence or dictionary (or list vs map, depending on your terminology) is itself a design flaw that should never be considered.  If you don't know whether a value is a list or a map, then you don't know jack about your data model.  PHP's mixing of the two into a single array type is on the level of `register_globals` in terms of how bad an idea it was, and we are still paying the price for it.

The Serde attributes aren't designed for generic references, so sue, they can be expanded.  My core point is that if I, as a library author inspecting the attributes, need to parse a string, then the design has failed and I have no interest in it.

**Comment by azjezz** [comment]
> The Serde attributes aren't designed for generic references, so sue, they can be expanded. My core point is that if I, as a library author inspecting the attributes, need to parse a string, then the design has failed and I have no interest in it.

I understand the desire to avoid string parsing. However, for a feature as complex and expressive as generics, there is no way to design a useful system within attributes that *doesn't* require parsing and type resolution. Simple attributes like `#[Dictionary(key: 'string', value: User::class)]` work for concrete types, but they break down immediately when you introduce template parameters, constraints, or unions.

The information we need to convey is inherently linguistic. You need a syntax to define templates, apply constraints, and reference those templates elsewhere. This requires a parser, whether it's processing a PHPDoc comment or an attribute's string argument. The collection example I gave was just a random example to question how it would even work. Here is a more realistic, non-collection example:

```php
<?php

/**
 * @template TId of IdentifierInterface
 * @template TEntity of object
 */
interface DatabaseRepository {
    /**
     * @param TId $id
     * @return TEntity|null
     */
    public function find(IdentifierInterface $id): ?object;
}
```

How would this be expressed with simple, pre-parsed attributes? You need to define two template parameters (`TId`, `TEntity`), apply constraints to them (`of IdentifierInterface`, `of object`), and then reference them in method signatures. This entire system of definition and reference is what static analysis tools handle, and it fundamentally relies on parsing a defined grammar. You can't escape the "string parsing" problem; you can only change where the string lives (PHPDoc vs. attribute).

And I'm not disagreeing that array in PHP is bad. I am 100% in support of splitting it into `vec`/`vector` and `dict`/`dictionary` types, but that is irrelevant to the generics discussion. In fact, a robust generics system is what would make those new types truly powerful by allowing us to create, for example, a `vec<User>` or a `dict<UserId, Session>`. The two issues are related, but generics is the more foundational feature needed to properly solve the typed collection problem.

**Comment by DaveLiddament** [comment]
This proposal for creating a standard for generics in PHP was written at the start of 2021. Until today I'd forgotten about this repo. Initially the purpose of this was to ask 2 questions:

1. Do we need to formalise the standard?
2. Should we use attributes rather the docblocks?

Given the way the past 4.5 years have played out the answer to both questions is, in my opinion, no. 

Longer answers to each question...

1. There was no need to formalise the standard; the SA tool maintainers have generally agreed on most points and the tools are largely compatible. It's worth noting that premature standardisation, which I was advocating for, would have hindered innovation. So I'm pleased no one listen to me! 

As a side note: If you are creating application code (which most devs are) and you find an odd edge case that works in one SA tool and not another, then that's a good signal to simplify the application code.  

2. Docblocks are still the least worst option available at the moment.  I flip flop between whether ideas like [this](https://gist.github.com/DaveLiddament/40130a7a107478bf6f92fcbb0b01a2fc) are a good or not. Currently I think not, as it'll slow down progress.

Thanks @PatchRanger for your interest in the proposal today, and thanks to the others who've offered their useful insights as to why my musings from almost half a decade ago were not a good idea.

**Comment by SerheyDolgushev** [comment]
👋 This time update in this thread comes almost 4 months after the previous message was posted. Which seems to be a great progress compared to 4 years 😄 


> 1. Do we need to formalise the standard?

Tagging @rlerdorf @Daimona and @dasl- here as this topic recently popped up in the context of https://github.com/phan/phan/wiki/Generic-Types.

> 2. Should we use attributes rather the docblocks?

Seems like one of the main benefits of moving to attributes is to avoid parsing strings. This could lead to performance improvements, but in the case of the Static Analyzer, it might not be really noticeable. As those strings are parsed on rare non-runtime occasions (which seems to be different for serializations).

On the other hand, introducing attributes will increase the size of the AST tree, and that might be noticeable on a huge codebase with wide utilization of those new attributes. And moving to the "attributes only" approach will make it impossible to use generics in legacy codebases. Which might not be a real problem, however, it is worth considering.
